NIFI-7888 Add support for SAML authentication

bbende opened the pull request:
This PR adds the ability to authenticate to NiFi via a SAML identity provider, similar to how OIDC authentication works.

In addition, there is an option to obtain group membership info from the assertions in a successful SAML AuthN response from the IDP. These groups are then passed along in the NiFiUser instance in order to be leveraged during authorization. Currently if using the file-based policy provider, then these groups would also have to exist in the configured user-group-provider in order to have created policies against them.

The integration with spring-security-saml is heavily based on the primary example application here:
https://github.com/vdenotaris/spring-boot-security-saml-sample

I've primarily tested against KeyCloak and the SSOCirlce IDP which is used by the example app above (https://www.ssocircle.com/en/). 

High-level changes:
- Add dependency on spring-security-saml2-core
- Updated AccessResource with new SAML end-points
- Updated Login/Logout filters to handle SAML scenario
- Updated logout process to track a logout request using a cookie
- Added database storage for cached SAML credential and user groups
- Updated proxied requests when clustered to send IDP groups in a header
- Updated X509 filter to process the IDP groups from the header if present
- Updated admin guide
- Fixed logout action on error page
- Updated StandardManagedAuthorizer to combine groups from request with groups from lookup
- Updated UserGroupProvider implementations with more efficient impl of getGroupByName
- Added/updated unit tests

mcgilman commented:
Thanks for the PR @bbende! Will review...

mcgilman commented on pom.xml:
```diff
@@ -139,6 +139,11 @@
+        <repository>
+            <id>Shibboleth</id>
+            <name>Shibboleth</name>
+            <url>https://build.shibboleth.net/nexus/content/repositories/releases/</url>
```
What is requiring this new repository? I see there is a new dependency for `spring-security-saml2-core` but that appears to be in maven central.

bbende replied:
Shibboleth is the repo where OpenSAML is released to and it seems like newer OpenSAML released are only in their specific repo and not in Maven central. 

The spring-security-saml project had to do something similar here:
https://github.com/spring-projects/spring-security-saml/blob/master/build.gradle#L50

mcgilman replied:
Thanks for the clarification.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/AccessResource.java:
```diff
@@ -171,6 +204,516 @@ public Response getLoginConfig(@Context HttpServletRequest httpServletRequest) {
+        httpServletResponse.sendRedirect(getNiFiUri());
+    }
+
+    private void storeIdpGroups(final String userIdentity, final IdpType idpType, final Set<String> groupNames) {
```
nit-pick... Should the logic in `samlLoginConsumer`, `storeIdpGroups`, and `samlSingleLogoutConsumer` be elsewhere? Seems like a lot to have in the `AccessResource`.

bbende replied:
I'll see if it makes sense to push some of this logic down into the service.

mcgilman commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/AccessResource.java:
```diff
@@ -171,6 +204,516 @@ public Response getLoginConfig(@Context HttpServletRequest httpServletRequest) {
+            logger.info("{} belongs to {}", userIdentity, mappedGroupName);
+        }
+
+        idpUserGroupService.createUserGroups(idpUserGroups);
```
Does `deleteUserGroups` and `createUserGroups` need tobe atomic?

bbende replied:
Good question. I thought about this at one point and never came back to it. I suppose it might make make more sense to have a `replaceUserGroups` that could handle all of the logic within a single transaction.

In practice I'm not sure how much of a problem it would really be if the delete worked and then create failed, the failure on insert would fail the whole login process and they'd have to try again anyway.

I can work on this change though, I think its the more correct approach to be atomic.

mcgilman reviewed (commented):
Code general looks great! Good test coverage. I've left a few minor comments. Still need to spend some time doing functional testing before I'm a +1 but this is an awesome contribution. Thanks @bbende!

thenatog commented:
I've also been reviewing this. I was able to set up Google Suite as a SAML IDP and log in. I am still testing a few things such as signing validation.

bbende commented:
@mcgilman @thenatog added some additional commits to address some of the review feedback and improve a few things I ran into while testing, here is a summary of the changes...

- Refactored some of the DB operations to have "replace" methods instead of calling "delete" and "create" in separate transactions where one could succeed and the second could fail

- Added properties for configuring the values of `AuthnRequestsSigned` and `WantAssertionsSigned` for the service provider metadata that is generated for nifi at /nifi-api/access/saml/metadata
    ```
    nifi.security.user.saml.request.signing.enabled=false
    nifi.security.user.saml.want.assertions.signed=true
    ```
- Remove the property for specifying the signing key alias, it now inspects the keystore and finds the private key entry and gets the alias automatically, if more than one private key entry exists then an exception is thrown (nifi already assumes a single private key in the keystore)

- Added a property for specifying an attribute to obtain the user identity from, if an attribute is not specified or if the attribute is not found in the response, then the Subject NameID is used by default
    `nifi.security.user.saml.identity.attribute.name=`

bbende commented on nifi-docs/src/main/asciidoc/administration-guide.adoc:
```diff
@@ -375,12 +375,32 @@ JSON Web Key (JWK) provided through the jwks_uri in the metadata found at the di
+|`nifi.security.user.saml.idp.metadata.url` | The URL for obtaining the identity provider's metadata. The metadata can be retrieved from the identity provider via `http://` or `https://`, or a local file can be referenced using `file://` .
+|`nifi.security.user.saml.sp.entity.id`| The entity id of the service provider (i.e. NiFi). This value will be used as the `Issuer` for SAML authentication requests and should be a valid URI. In some cases the service provider entity id must be registered ahead of time with the identity provider.
+|`nifi.security.user.saml.signing.key.alias`| The alias of the key within `nifi.security.keystore` that will be used for signing SAML messages.
+|`nifi.security.user.saml.signature.algorithm`| The algorithm to use when signing SAML messages. Reference the link:https://git.shibboleth.net/view/?p=java-xmltooling.git;a=blob;f=src/main/java/org/opensaml/xml/signature/SignatureConstants.java[Open SAML Signature Constants] for a list of valid values. If not specified the default of SHA-1 will be used.
```
What I meant when I wrote this was that if we don't pass in any value to Open SAML then they default to SHA-1. After writing these descriptions I updated the default nifi.properties to have SHA-256. So the description basically means if you delete the default value in nifi.properties so that its empty, then SHA-1 gets used. I could take it a step further and update NiFiProperties itself to have a default as well.

bbende commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/saml/impl/StandardSAMLConfigurationFactory.java:
```diff
@@ -0,0 +1,502 @@
+        // Consider not using global registration of protocol here as it would potentially impact other uses of commons http client
+        // with in nifi-framework-nar, currently there are no other usages, see https://hc.apache.org/httpclient-3.x/sslguide.html
+        final Protocol p = new Protocol("https", socketFactory, 443);
+        Protocol.registerProtocol(p.getScheme(), p);
```
Correct, Protocol is a class specific to commons http client, and this is registering the socket factory so that any commons http clients that make https connections will use the custom socket factory.

bbende commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/saml/impl/tls/CustomTLSProtocolSocketFactory.java:
```diff
@@ -0,0 +1,69 @@
+import java.net.Socket;
+import java.net.SocketAddress;
+
+public class CustomTLSProtocolSocketFactory implements SecureProtocolSocketFactory {
```
Correct, this class is used when the "truststore strategy" is "NIFI" so that an SSLContext from NiFI's TLS properties will be used. Otherwise it defaults to the JDK system truststore.

bbende commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/saml/impl/tls/CompositeKeyManager.java:
```diff
@@ -0,0 +1,107 @@
+ *
+ * The default Credential is considered that default Credential from the server KeyManager.
+ */
+public class CompositeKeyManager implements KeyManager {
```
Correct, spring-security-saml is based around a single keystore that contains private key and trusted certs, so I made this wrapper to combine the two separate keystores.

mcgilman commented:
@bbende Thanks again for the PR! My earlier concerns have been addressed and things look good. Was able to verify the functionality. I am a +1 once the comments from @thenatog are addressed.

bbende commented:
Thanks @mcgilman  !

@thenatog I pushed up another commit that makes SHA-256 the default when no value is provided in nifi.properties for the signature or digest algorithms. Let me know if there is anything else needed to get this merged, thanks!
